### DISCUSSION TITLE
Fatal Error: Can't open module file 'mpi.mod' for reading at (1): No such file or directory

### DISCUSSION THREAD

**Discussion by agetiran** [NASA-LIS/LISF#836] [Q&A]
Hi,
I've been trying to compile LIS on a server, but get this error message at the compilation of the first routine using mpi:

Fatal Error: Can't open module file 'mpi.mod' for reading at (1): No such file or directory

I have OpenMPI installed and its info is provided at the end of this message. The lines below are used in the .bashrc file:

export PATH=$PATH:$HOME/opt/openmpi/bin
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$HOME/opt/openmpi/lib

It seems LIS is looking for mpi at the wrong place. Does anyone have know how to fix this or where I should start looking? It's my first time compiling LIS outside Discover and I'm a bit lost.

Thanks,
Augusto

[root@SRV000443 ~]# ompi_info
                 Package: Open MPI root@SRV000443 Distribution
                Open MPI: 4.1.1
  Open MPI repo revision: v4.1.1
   Open MPI release date: Apr 24, 2021
                Open RTE: 4.1.1
  Open RTE repo revision: v4.1.1
   Open RTE release date: Apr 24, 2021
                    OPAL: 4.1.1
      OPAL repo revision: v4.1.1
       OPAL release date: Apr 24, 2021
                 MPI API: 3.1.0
            Ident string: 4.1.1
                  Prefix: /root/opt/openmpi
 Configured architecture: x86_64-pc-linux-gnu
          Configure host: SRV000443
           Configured by: root
           Configured on: Thu Jul  8 00:52:54 UTC 2021
          Configure host: SRV000443
  Configure command line: '--prefix=/root/opt/openmpi'
                Built by: root
                Built on: Thu Jul  8 01:16:23 UTC 2021
              Built host: SRV000443
              C bindings: yes
            C++ bindings: no
             Fort mpif.h: yes (all)
            Fort use mpi: yes (limited: overloading)
       Fort use mpi size: deprecated-ompi-info-value
        Fort use mpi_f08: no
 Fort mpi_f08 compliance: The mpi_f08 module was not built
  Fort mpi_f08 subarrays: no
           Java bindings: no
  Wrapper compiler rpath: runpath
              C compiler: gcc
     C compiler absolute: /usr/bin/gcc
  C compiler family name: GNU
      C compiler version: 4.8.5
            C++ compiler: g++
   C++ compiler absolute: /usr/bin/g++
           Fort compiler: gfortran
       Fort compiler abs: /usr/bin/gfortran
         Fort ignore TKR: no
   Fort 08 assumed shape: no
      Fort optional args: no
          Fort INTERFACE: yes
    Fort ISO_FORTRAN_ENV: yes
       Fort STORAGE_SIZE: no
      Fort BIND(C) (all): no
      Fort ISO_C_BINDING: yes
 Fort SUBROUTINE BIND(C): no
       Fort TYPE,BIND(C): no
 Fort T,BIND(C,name="a"): no
            Fort PRIVATE: no
          Fort PROTECTED: no
           Fort ABSTRACT: no
       Fort ASYNCHRONOUS: no
          Fort PROCEDURE: no
         Fort USE...ONLY: no
           Fort C_FUNLOC: no
 Fort f08 using wrappers: no
         Fort MPI_SIZEOF: no
             C profiling: yes
           C++ profiling: no
   Fort mpif.h profiling: yes
  Fort use mpi profiling: yes
   Fort use mpi_f08 prof: no
          C++ exceptions: no
          Thread support: posix (MPI_THREAD_MULTIPLE: yes, OPAL support: yes,
                          OMPI progress: no, ORTE progress: yes, Event lib:
                          yes)
           Sparse Groups: no
  Internal debug support: no
  MPI interface warnings: yes
     MPI parameter check: runtime
Memory profiling support: no
Memory debugging support: no
              dl support: yes
   Heterogeneous support: no
 mpirun default --prefix: no
       MPI_WTIME support: native
     Symbol vis. support: yes
   Host topology support: yes
            IPv6 support: no
      MPI1 compatibility: no
          MPI extensions: affinity, cuda, pcollreq
   FT Checkpoint support: no (checkpoint thread: no)
   C/R Enabled Debugging: no
  MPI_MAX_PROCESSOR_NAME: 256
    MPI_MAX_ERROR_STRING: 256
     MPI_MAX_OBJECT_NAME: 64
        MPI_MAX_INFO_KEY: 36
        MPI_MAX_INFO_VAL: 256
       MPI_MAX_PORT_NAME: 1024
  MPI_MAX_DATAREP_STRING: 128
           MCA allocator: basic (MCA v2.1.0, API v2.0.0, Component v4.1.1)
           MCA allocator: bucket (MCA v2.1.0, API v2.0.0, Component v4.1.1)
           MCA backtrace: execinfo (MCA v2.1.0, API v2.0.0, Component v4.1.1)
                 MCA btl: self (MCA v2.1.0, API v3.1.0, Component v4.1.1)
                 MCA btl: tcp (MCA v2.1.0, API v3.1.0, Component v4.1.1)
                 MCA btl: vader (MCA v2.1.0, API v3.1.0, Component v4.1.1)
            MCA compress: bzip (MCA v2.1.0, API v2.0.0, Component v4.1.1)
            MCA compress: gzip (MCA v2.1.0, API v2.0.0, Component v4.1.1)
                 MCA crs: none (MCA v2.1.0, API v2.0.0, Component v4.1.1)
                  MCA dl: dlopen (MCA v2.1.0, API v1.0.0, Component v4.1.1)
               MCA event: libevent2022 (MCA v2.1.0, API v2.0.0, Component
                          v4.1.1)
               MCA hwloc: hwloc201 (MCA v2.1.0, API v2.0.0, Component v4.1.1)
                  MCA if: linux_ipv6 (MCA v2.1.0, API v2.0.0, Component
                          v4.1.1)
                  MCA if: posix_ipv4 (MCA v2.1.0, API v2.0.0, Component
                          v4.1.1)
         MCA installdirs: env (MCA v2.1.0, API v2.0.0, Component v4.1.1)
         MCA installdirs: config (MCA v2.1.0, API v2.0.0, Component v4.1.1)
              MCA memory: patcher (MCA v2.1.0, API v2.0.0, Component v4.1.1)
               MCA mpool: hugepage (MCA v2.1.0, API v3.0.0, Component v4.1.1)
             MCA patcher: overwrite (MCA v2.1.0, API v1.0.0, Component
                          v4.1.1)
                MCA pmix: isolated (MCA v2.1.0, API v2.0.0, Component v4.1.1)
                MCA pmix: flux (MCA v2.1.0, API v2.0.0, Component v4.1.1)
                MCA pmix: pmix3x (MCA v2.1.0, API v2.0.0, Component v4.1.1)
               MCA pstat: linux (MCA v2.1.0, API v2.0.0, Component v4.1.1)
              MCA rcache: grdma (MCA v2.1.0, API v3.3.0, Component v4.1.1)
           MCA reachable: weighted (MCA v2.1.0, API v2.0.0, Component v4.1.1)
               MCA shmem: mmap (MCA v2.1.0, API v2.0.0, Component v4.1.1)
               MCA shmem: posix (MCA v2.1.0, API v2.0.0, Component v4.1.1)
               MCA shmem: sysv (MCA v2.1.0, API v2.0.0, Component v4.1.1)
               MCA timer: linux (MCA v2.1.0, API v2.0.0, Component v4.1.1)
              MCA errmgr: default_app (MCA v2.1.0, API v3.0.0, Component
                          v4.1.1)
              MCA errmgr: default_hnp (MCA v2.1.0, API v3.0.0, Component
                          v4.1.1)
              MCA errmgr: default_orted (MCA v2.1.0, API v3.0.0, Component
                          v4.1.1)
              MCA errmgr: default_tool (MCA v2.1.0, API v3.0.0, Component
                          v4.1.1)
                 MCA ess: env (MCA v2.1.0, API v3.0.0, Component v4.1.1)
                 MCA ess: hnp (MCA v2.1.0, API v3.0.0, Component v4.1.1)
                 MCA ess: pmi (MCA v2.1.0, API v3.0.0, Component v4.1.1)
                 MCA ess: singleton (MCA v2.1.0, API v3.0.0, Component
                          v4.1.1)
                 MCA ess: tool (MCA v2.1.0, API v3.0.0, Component v4.1.1)
                 MCA ess: slurm (MCA v2.1.0, API v3.0.0, Component v4.1.1)
               MCA filem: raw (MCA v2.1.0, API v2.0.0, Component v4.1.1)
             MCA grpcomm: direct (MCA v2.1.0, API v3.0.0, Component v4.1.1)
                 MCA iof: hnp (MCA v2.1.0, API v2.0.0, Component v4.1.1)
                 MCA iof: orted (MCA v2.1.0, API v2.0.0, Component v4.1.1)
                 MCA iof: tool (MCA v2.1.0, API v2.0.0, Component v4.1.1)
                MCA odls: default (MCA v2.1.0, API v2.0.0, Component v4.1.1)
                MCA odls: pspawn (MCA v2.1.0, API v2.0.0, Component v4.1.1)
                 MCA oob: tcp (MCA v2.1.0, API v2.0.0, Component v4.1.1)
                 MCA plm: isolated (MCA v2.1.0, API v2.0.0, Component v4.1.1)
                 MCA plm: rsh (MCA v2.1.0, API v2.0.0, Component v4.1.1)
                 MCA plm: slurm (MCA v2.1.0, API v2.0.0, Component v4.1.1)
                 MCA ras: simulator (MCA v2.1.0, API v2.0.0, Component
                          v4.1.1)
                 MCA ras: slurm (MCA v2.1.0, API v2.0.0, Component v4.1.1)
                MCA regx: fwd (MCA v2.1.0, API v1.0.0, Component v4.1.1)
                MCA regx: naive (MCA v2.1.0, API v1.0.0, Component v4.1.1)
                MCA regx: reverse (MCA v2.1.0, API v1.0.0, Component v4.1.1)
               MCA rmaps: mindist (MCA v2.1.0, API v2.0.0, Component v4.1.1)
               MCA rmaps: ppr (MCA v2.1.0, API v2.0.0, Component v4.1.1)
               MCA rmaps: rank_file (MCA v2.1.0, API v2.0.0, Component
                          v4.1.1)
               MCA rmaps: resilient (MCA v2.1.0, API v2.0.0, Component
                          v4.1.1)
               MCA rmaps: round_robin (MCA v2.1.0, API v2.0.0, Component
                          v4.1.1)
               MCA rmaps: seq (MCA v2.1.0, API v2.0.0, Component v4.1.1)
                 MCA rml: oob (MCA v2.1.0, API v3.0.0, Component v4.1.1)
              MCA routed: binomial (MCA v2.1.0, API v3.0.0, Component v4.1.1)
              MCA routed: direct (MCA v2.1.0, API v3.0.0, Component v4.1.1)
              MCA routed: radix (MCA v2.1.0, API v3.0.0, Component v4.1.1)
                 MCA rtc: hwloc (MCA v2.1.0, API v1.0.0, Component v4.1.1)
              MCA schizo: flux (MCA v2.1.0, API v1.0.0, Component v4.1.1)
              MCA schizo: ompi (MCA v2.1.0, API v1.0.0, Component v4.1.1)
              MCA schizo: orte (MCA v2.1.0, API v1.0.0, Component v4.1.1)
              MCA schizo: jsm (MCA v2.1.0, API v1.0.0, Component v4.1.1)
              MCA schizo: slurm (MCA v2.1.0, API v1.0.0, Component v4.1.1)
               MCA state: app (MCA v2.1.0, API v1.0.0, Component v4.1.1)
               MCA state: hnp (MCA v2.1.0, API v1.0.0, Component v4.1.1)
               MCA state: novm (MCA v2.1.0, API v1.0.0, Component v4.1.1)
               MCA state: orted (MCA v2.1.0, API v1.0.0, Component v4.1.1)
               MCA state: tool (MCA v2.1.0, API v1.0.0, Component v4.1.1)
                 MCA bml: r2 (MCA v2.1.0, API v2.0.0, Component v4.1.1)
                MCA coll: adapt (MCA v2.1.0, API v2.0.0, Component v4.1.1)
                MCA coll: basic (MCA v2.1.0, API v2.0.0, Component v4.1.1)
                MCA coll: han (MCA v2.1.0, API v2.0.0, Component v4.1.1)
                MCA coll: inter (MCA v2.1.0, API v2.0.0, Component v4.1.1)
                MCA coll: libnbc (MCA v2.1.0, API v2.0.0, Component v4.1.1)
                MCA coll: self (MCA v2.1.0, API v2.0.0, Component v4.1.1)
                MCA coll: sm (MCA v2.1.0, API v2.0.0, Component v4.1.1)
                MCA coll: sync (MCA v2.1.0, API v2.0.0, Component v4.1.1)
                MCA coll: tuned (MCA v2.1.0, API v2.0.0, Component v4.1.1)
                MCA coll: monitoring (MCA v2.1.0, API v2.0.0, Component
                          v4.1.1)
                MCA fbtl: posix (MCA v2.1.0, API v2.0.0, Component v4.1.1)
               MCA fcoll: dynamic (MCA v2.1.0, API v2.0.0, Component v4.1.1)
               MCA fcoll: dynamic_gen2 (MCA v2.1.0, API v2.0.0, Component
                          v4.1.1)
               MCA fcoll: individual (MCA v2.1.0, API v2.0.0, Component
                          v4.1.1)
               MCA fcoll: two_phase (MCA v2.1.0, API v2.0.0, Component
                          v4.1.1)
               MCA fcoll: vulcan (MCA v2.1.0, API v2.0.0, Component v4.1.1)
                  MCA fs: ufs (MCA v2.1.0, API v2.0.0, Component v4.1.1)
                  MCA io: ompio (MCA v2.1.0, API v2.0.0, Component v4.1.1)
                  MCA io: romio321 (MCA v2.1.0, API v2.0.0, Component v4.1.1)
                  MCA op: avx (MCA v2.1.0, API v1.0.0, Component v4.1.1)
                 MCA osc: sm (MCA v2.1.0, API v3.0.0, Component v4.1.1)
                 MCA osc: monitoring (MCA v2.1.0, API v3.0.0, Component
                          v4.1.1)
                 MCA osc: pt2pt (MCA v2.1.0, API v3.0.0, Component v4.1.1)
                 MCA osc: rdma (MCA v2.1.0, API v3.0.0, Component v4.1.1)
                 MCA pml: v (MCA v2.1.0, API v2.0.0, Component v4.1.1)
                 MCA pml: cm (MCA v2.1.0, API v2.0.0, Component v4.1.1)
                 MCA pml: monitoring (MCA v2.1.0, API v2.0.0, Component
                          v4.1.1)
                 MCA pml: ob1 (MCA v2.1.0, API v2.0.0, Component v4.1.1)
                 MCA rte: orte (MCA v2.1.0, API v2.0.0, Component v4.1.1)
            MCA sharedfp: individual (MCA v2.1.0, API v2.0.0, Component
                          v4.1.1)
            MCA sharedfp: lockedfile (MCA v2.1.0, API v2.0.0, Component
                          v4.1.1)
            MCA sharedfp: sm (MCA v2.1.0, API v2.0.0, Component v4.1.1)
                MCA topo: basic (MCA v2.1.0, API v2.2.0, Component v4.1.1)
                MCA topo: treematch (MCA v2.1.0, API v2.2.0, Component
                          v4.1.1)
           MCA vprotocol: pessimist (MCA v2.1.0, API v2.0.0, Component
                          v4.1.1)

**Comment by dmocko** [comment]
Hi Augusto,

I have not yet tried to use Open MPI - but I have a couple of quick suggestions:

1) The result of the "ompi_info" command lists the Prefix as: /root/opt/openmpi.  However, you are using $HOME/opt/openmpi/[bin or lib] in your $PATH and $LD_LIBRARY_PATH.  Have you tried adding /root/opt/openmpi/[bin or lib] instead?

2) I would suggest trying to create a custom modulefile.  It can ensure that all of the LIS environmental variables are defined and consistent with your configuration/compilation:
     https://github.com/NASA-LIS/LISF/blob/master/docs/creating_a_custom_modulefile/creating-a-custom-modulefile.adoc
If you can't create a custom modulefile, then refer to the build document to check that you have everything defined:
    https://github.com/NASA-LIS/LISF/blob/master/docs/LIS_users_guide/build.adoc

David

**Comment by agetiran** [comment]
Hi David,
Thanks for the suggestions and for catching these inconsistencies. Unfortunately, these fixes haven’t eliminated the error message.

Augusto


From: David Mocko ***@***.***>
Reply-To: NASA-LIS/LISF ***@***.***>
Date: Thursday, July 15, 2021 at 4:30 PM
To: NASA-LIS/LISF ***@***.***>
Cc: "Getirana, Augusto (GSFC-617.0)[Science Application International Corp.]" ***@***.***>, Author ***@***.***>
Subject: [EXTERNAL] Re: [NASA-LIS/LISF] Fatal Error: Can't open module file 'mpi.mod' for reading at (1): No such file or directory (#836)


Hi Augusto,

I have not yet tried to use Open MPI - but I have a couple of quick suggestions:

  1.  The result of the "ompi_info" command lists the Prefix as: /root/opt/openmpi. However, you are using $HOME/opt/openmpi/[bin or lib] in your $PATH and $LD_LIBRARY_PATH. Have you tried adding /root/opt/openmpi/[bin or lib] instead?
  2.  I would suggest trying to create a custom modulefile. It can ensure that all of the LIS environmental variables are defined and consistent with your configuration/compilation:
https://github.com/NASA-LIS/LISF/blob/master/docs/creating_a_custom_modulefile/creating-a-custom-modulefile.adoc<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FNASA-LIS%2FLISF%2Fblob%2Fmaster%2Fdocs%2Fcreating_a_custom_modulefile%2Fcreating-a-custom-modulefile.adoc&data=04%7C01%7Caugusto.getirana%40nasa.gov%7Ccfca04967f8c4f74c5cd08d947cf6841%7C7005d45845be48ae8140d43da96dd17b%7C0%7C0%7C637619778396914830%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=SL6orYM3jtBmwy6X%2B1ktJLyph1W5%2BgUJp2tJp95TPaA%3D&reserved=0>
If you can't create a custom modulefile, then refer to the build document to check that you have everything defined:
https://github.com/NASA-LIS/LISF/blob/master/docs/LIS_users_guide/build.adoc<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FNASA-LIS%2FLISF%2Fblob%2Fmaster%2Fdocs%2FLIS_users_guide%2Fbuild.adoc&data=04%7C01%7Caugusto.getirana%40nasa.gov%7Ccfca04967f8c4f74c5cd08d947cf6841%7C7005d45845be48ae8140d43da96dd17b%7C0%7C0%7C637619778396924785%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=oL%2F8tnFewFsIRCu3m6Y7f3N4fonTXKZzFx%2BNyqTC8V4%3D&reserved=0>

David

**Comment by jvgeiger** [comment]
@agetiran, may I close this discussion?

**Comment by hajun97** [comment]
Hi Augusto,
May I ask how you solved your problem? I am experiencing the same situation.
Jun

**Comment by jvgeiger** [comment]
Hello,

mpi.mod is provided by your MPI library package.  Your MPI compiler wrappers, also provided by your MPI library package, called mpif90 and mpicc, are supposed to automatically add all relevant MPI variables when compiling.


On my system, MPI is installed in /usr/local/intel/oneapi/2021/mpi/2021.4.0/.  Note the `-I` and `-L` variables that are set by mpif90.

```
% mpif90 -show | string split " "
/usr/local/intel/oneapi/2021/compiler/2021.4.0/linux/bin/intel64/ifort
-I"/usr/local/intel/oneapi/2021/mpi/2021.4.0/include/gfortran"
-I"/usr/local/intel/oneapi/2021/mpi/2021.4.0/include"
-L"/usr/local/intel/oneapi/2021/mpi/2021.4.0/lib/release"
-L"/usr/local/intel/oneapi/2021/mpi/2021.4.0/lib"
-Xlinker
--enable-new-dtags
-Xlinker
-rpath
-Xlinker
"/usr/local/intel/oneapi/2021/mpi/2021.4.0/lib/release"
-Xlinker
-rpath
-Xlinker
"/usr/local/intel/oneapi/2021/mpi/2021.4.0/lib"
-lmpifort
-lmpi
-lrt
-lpthread
-Wl,-z,now
-Wl,-z,relro
-Wl,-z,noexecstack
-Xlinker
--enable-new-dtags
-ldl
```


You must make sure that your MPI library package was installed correctly and that your MPI wrappers contain the correct settings.
